main/lsp/requests/workspace_symbols.cc

```diff
@@ -165,101 +229,89 @@ vector<unique_ptr<SymbolInformation>> SymbolMatcher::doQuery(string_view query_v
+                    continue;
                 }
-                if (ancestorEnd == queryEnd) {
-                    continue; // ancestor matched everything, so skip to its parent
+                auto shortName = nameData->shortName(gs);
+                uint bestScore = ceilingScore;
+                auto [partialScore, partialMatchEnd] =
+                    partialMatchSymbol(shortName, queryBegin, queryEnd, false, ceilingScore);
+                if (partialMatchEnd == queryEnd) {
+                    bestScore = partialScore;
                 }
-                if (ancestorScore >= bestScore) {
-                    continue; // matching this ancestor would be worse
+                for (auto previousAncestorRef = symbolRef, ancestorRef = symbolData->owner;
+                     previousAncestorRef != ancestorRef && ancestorRef.exists();
+                     previousAncestorRef = ancestorRef, ancestorRef = ancestorRef.data(gs)->owner) {
+                    auto [ancestorScore, ancestorEnd] = getPartialMatch(ancestorRef);
+                    if (ancestorEnd == queryBegin || ancestorEnd == nullptr) {
+                        break; // no further ancestor will be of any help
+                    }
+                    if (ancestorEnd == queryEnd) {
+                        continue; // ancestor matched everything, so skip to its parent
+                    }
+                    if (ancestorScore >= bestScore) {
+                        continue; // matching this ancestor would be worse
+                    }
+                    auto [plusScore, plusEnd] =
+                        partialMatchSymbol(shortName, ancestorEnd, queryEnd, false, bestScore - ancestorScore);
+                    if (plusEnd == queryEnd) {
+                        bestScore = min(bestScore, ancestorScore + plusScore);
+                    }
                 }
-                auto [plusScore, plusEnd] =
-                    partialMatchSymbol(shortName, ancestorEnd, queryEnd, false, bestScore - ancestorScore);
-                if (plusEnd == queryEnd) {
-                    bestScore = min(bestScore, ancestorScore + plusScore);
+                if (bestScore < ceilingScore) {
+                    // update ceiling so we stop looking at bad matches
+                    ceilingScore = min(ceilingScore, bestScore * WORST_TO_BEST_RATIO);
+                    candidates.emplace_back(symbolRef, bestScore);
                 }
             }
-            if (bestScore < ceilingScore) {
-                // update ceiling so we stop looking at bad matches
-                ceilingScore = min(ceilingScore, bestScore * WORST_TO_BEST_RATIO);
-                candidates.emplace_back(symbolIndex, bestScore);
-            }
         }
     }
-    fast_sort(candidates, [](pair<u4, int> &left, pair<u4, int> &right) -> bool { return left.second < right.second; });
+    fast_sort(candidates, [](pair<core::SymbolRef, int> &left, pair<core::SymbolRef, int> &right) -> bool {
```

Comment:
Do you need the argument types here?
```suggestion
    fast_sort(candidates, [](auto &left, auto &right) -> bool {
```

Reply:
let's find out